In [9]:
#!pip install ucimlrepo

import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns
from sklearn.preprocessing import LabelEncoder

import warnings
warnings.filterwarnings('ignore')

In [1]:
from ucimlrepo import fetch_ucirepo 
  
# fetch dataset 
adult = fetch_ucirepo(id=2) 
  
# data (as pandas dataframes) 
X = adult.data.features 
y = adult.data.targets 
  
# metadata 
print(adult.metadata) 
  
# variable information 
print(adult.variables) 

{'uci_id': 2, 'name': 'Adult', 'repository_url': 'https://archive.ics.uci.edu/dataset/2/adult', 'data_url': 'https://archive.ics.uci.edu/static/public/2/data.csv', 'abstract': 'Predict whether annual income of an individual exceeds $50K/yr based on census data. Also known as "Census Income" dataset. ', 'area': 'Social Science', 'tasks': ['Classification'], 'characteristics': ['Multivariate'], 'num_instances': 48842, 'num_features': 14, 'feature_types': ['Categorical', 'Integer'], 'demographics': ['Age', 'Income', 'Education Level', 'Other', 'Race', 'Sex'], 'target_col': ['income'], 'index_col': None, 'has_missing_values': 'yes', 'missing_values_symbol': 'NaN', 'year_of_dataset_creation': 1996, 'last_updated': 'Tue Sep 24 2024', 'dataset_doi': '10.24432/C5XW20', 'creators': ['Barry Becker', 'Ronny Kohavi'], 'intro_paper': None, 'additional_info': {'summary': "Extraction was done by Barry Becker from the 1994 Census database.  A set of reasonably clean records was extracted using the fol

In [7]:
# Fetching the required dataset from list of UCI repo

adult = fetch_ucirepo(id=2)




In [10]:
# Creating dataframe with feature columns and target columns

x=pd.DataFrame(adultdataset.data.features)

y=pd.DataFrame(adultdataset.data.targets,columns=['income'])

df=pd.concat([x,y],axis=1)   # To merge this 2 datasets

df.to_excel("adultdata.xlsx",index=False)   # To save this dataframe to excel into local machine

In [25]:
df.head(20)

age         workclass  fnlwgt     education  education-num  \
0    39         State-gov   77516     Bachelors             13   
1    50  Self-emp-not-inc   83311     Bachelors             13   
2    38           Private  215646       HS-grad              9   
3    53           Private  234721          11th              7   
4    28           Private  338409     Bachelors             13   
5    37           Private  284582       Masters             14   
6    49           Private  160187           9th              5   
7    52  Self-emp-not-inc  209642       HS-grad              9   
8    31           Private   45781       Masters             14   
9    42           Private  159449     Bachelors             13   
10   37           Private  280464  Some-college             10   
11   30         State-gov  141297     Bachelors             13   
12   23           Private  122272     Bachelors             13   
13   32           Private  205019    Assoc-acdm             12   
14   40           Private  121772     Assoc-voc             11   
15   34           Private  245487       7th-8th              4   
16   25  Self-emp-not-inc  176756       HS-grad              9   
17   32           Private  186824       HS-grad              9   
18   38           Private   28887          11th              7   
19   43  Self-emp-not-inc  292175       Masters             14   

           marital-status         occupation   relationship  \
0           Never-married       Adm-clerical  Not-in-family   
1      Married-civ-spouse    Exec-managerial        Husband   
2                Divorced  Handlers-cleaners  Not-in-family   
3      Married-civ-spouse  Handlers-cleaners        Husband   
4      Married-civ-spouse     Prof-specialty           Wife   
5      Married-civ-spouse    Exec-managerial           Wife   
6   Married-spouse-absent      Other-service  Not-in-family   
7      Married-civ-spouse    Exec-managerial        Husband   
8           Never-married     Prof-specialty  Not-in-family   
9      Married-civ-spouse    Exec-managerial        Husband   
10     Married-civ-spouse    Exec-managerial        Husband   
11     Married-civ-spouse     Prof-specialty        Husband   
12          Never-married       Adm-clerical      Own-child   
13          Never-married              Sales  Not-in-family   
14     Married-civ-spouse       Craft-repair        Husband   
15     Married-civ-spouse   Transport-moving        Husband   
16          Never-married    Farming-fishing      Own-child   
17          Never-married  Machine-op-inspct      Unmarried   
18     Married-civ-spouse              Sales        Husband   
19               Divorced    Exec-managerial      Unmarried   

                  race     sex  capital-gain  capital-loss  hours-per-week  \
0                White    Male          2174             0              40   
1                White    Male             0             0              13   
2                White    Male             0             0              40   
3                Black    Male             0             0              40   
4                Black  Female             0             0              40   
5                White  Female             0             0              40   
6                Black  Female             0             0              16   
7                White    Male             0             0              45   
8                White  Female         14084             0              50   
9                White    Male          5178             0              40   
10               Black    Male             0             0              80   
11  Asian-Pac-Islander    Male             0             0              40   
12               White  Female             0             0              30   
13               Black    Male             0             0              50   
14  Asian-Pac-Islander    Male             0             0              40   
15  Amer-Indian-Eskimo    Male             0 

<h1 style ="color: #200FBD; font-weight: bold; font-size: 32px; font-family: Baskerville Old Face"> Analyzing data frame features </h1>

In [39]:
# df.describe   # describes descriptive statistics for numerical columns by default. 
df.describe().T  # summary statistics of describe in transpose view

count           mean            std      min       25%  \
age             48842.0      38.643585      13.710510     17.0      28.0   
fnlwgt          48842.0  189664.134597  105604.025423  12285.0  117550.5   
education-num   48842.0      10.078089       2.570973      1.0       9.0   
capital-gain    48842.0    1079.067626    7452.019058      0.0       0.0   
capital-loss    48842.0      87.502314     403.004552      0.0       0.0   
hours-per-week  48842.0      40.422382      12.391444      1.0      40.0   

                     50%       75%        max  
age                 37.0      48.0       90.0  
fnlwgt          178144.5  237642.0  1490400.0  
education-num       10.0      12.0       16.0  
capital-gain         0.0       0.0    99999.0  
capital-loss         0.0       0.0     4356.0  
hours-per-week      40.0      45.0       99.0

In [42]:
df.shape


(48842, 15)

In [54]:

#df.dtypes  #overview of each column's data types 

numerical_cols = df.select_dtypes(include=['int64']).columns.tolist()
numerical_cols


['age',
 'fnlwgt',
 'education-num',
 'capital-gain',
 'capital-loss',
 'hours-per-week']

In [53]:
categorical_cols = df.select_dtypes(include=['object']).columns.tolist()
categorical_cols

['workclass',
 'education',
 'marital-status',
 'occupation',
 'relationship',
 'race',
 'sex',
 'native-country',
 'income']

<h3>Notes: Observations on Binary classification Problem</h3>
<span style =" color: #200FBD; font-size: 15px; font-family: Baskerville Old Face; font-weight: bold;">
<ol>
  <li>Most of the values in the dataset are <strong>categorical variables</strong>.</li>
  <li>The <strong>target variable is income</strong>, which makes this a <strong>binary classification problem</strong>.
    <ul>
      <li>The goal is to train a model to predict income class (low or high).</li>
      <li>There are only 2 possible outcomes, so it qualifies as binary classification.</li>
    </ul>
  </li>
  <li><strong>Features vs Target</strong>:
    <ul>
      <li>Features (inputs):  age, hours-per-week, etc.</li>
      <li>Target (output):    income</li>
    </ul>
  </li>
</ol>


In [60]:
df_sorted_age = df.sort_values(by='age', ascending=True)
df_sorted_age.tail()

age  workclass  fnlwgt     education  education-num  \
5104    90    Private   52386  Some-college             10   
18725   90  Local-gov  153602       HS-grad              9   
222     90    Private   51744       HS-grad              9   
31696   90          ?  313986       HS-grad              9   
6624    90    Private  313986          11th              7   

           marital-status     occupation   relationship                race  \
5104        Never-married  Other-service  Not-in-family  Asian-Pac-Islander   
18725  Married-civ-spouse  Other-service        Husband               White   
222         Never-married  Other-service  Not-in-family               Black   
31696  Married-civ-spouse              ?        Husband               White   
6624   Married-civ-spouse   Craft-repair        Husband               White   

        sex  capital-gain  capital-loss  hours-per-week native-country income  
5104   Male             0             0              35  United-States  <=50K  
18725  Male          6767             0              40  United-States  <=50K  
222    Male             0          2206              40  United-States  <=50K  
31696  Male             0             0              40  United-States   >50K  
6624   Male             0             0              40  United-States  <=50K

<h3>Notes: Observations on Case studies whose data is missing yet target income is >=50k </h3>
<span style =" color: #200FBD; font-size: 15px; font-family: Baskerville Old Face; font-weight: bold;">
<ol>
  <li>Record 31696 = workclass, occupation are '?' , capital-gain and capital-loss fields are '0'</strong>.</li>
  </ul>
  </li>
</ol>


### Data cleaning - Income 

In [62]:
df['income'].unique()

array(['<=50K', '>50K', '<=50K.', '>50K.'], dtype=object)

In [64]:
backup = df.copy()

In [65]:
df['income'] = df['income'].replace('<=50k.','<=50K')
df['income'] = df['income'].replace('>50K.', '>50K')

In [66]:
df['income'].value_counts()

income
<=50K     24720
<=50K.    12435
>50K      11687
Name: count, dtype: int64

In [67]:
df['income'] = df['income'].replace('<=50K.','<=50K')

In [68]:
df['income'].value_counts()

income
<=50K    37155
>50K     11687
Name: count, dtype: int64

### Identify similar records 

In [70]:
df[(df['workclass'] == '?') & (df['occupation'] == '?') & (df['income'] == '>50K')]

age workclass  fnlwgt     education  education-num      marital-status  \
27      54         ?  180211  Some-college             10  Married-civ-spouse   
154     52         ?  252903       HS-grad              9            Divorced   
354     59         ?  372020     Bachelors             13  Married-civ-spouse   
408     28         ?  123147  Some-college             10  Married-civ-spouse   
580     65         ?  195733  Some-college             10  Married-civ-spouse   
...    ...       ...     ...           ...            ...                 ...   
31997   60         ?  268954  Some-college             10  Married-civ-spouse   
31998   52         ?   89951          12th              8  Married-civ-spouse   
32016   62         ?  178764       HS-grad              9  Married-civ-spouse   
32530   35         ?  320084     Bachelors             13  Married-civ-spouse   
32539   71         ?  287372     Doctorate             16  Married-civ-spouse   

      occupation   relationship                race     sex  capital-gain  \
27             ?        Husband  Asian-Pac-Islander    Male             0   
154            ?  Not-in-family               White    Male             0   
354            ?        Husband               White    Male             0   
408            ?           Wife               White  Female             0   
580            ?        Husband               White    Male             0   
...          ...            ...                 ...     ...           ...   
31997          ?        Husband               White    Male             0   
31998          ?           Wife               Black  Female             0   
32016          ?        Husband               White    Male             0   
32530          ?           Wife               White  Female             0   
32539          ?        Husband               White    Male             0   

       capital-loss  hours-per-week native-country income  
27                0              60          South   >50K  
154               0              45  United-States   >50K  
354               0              40  United-States   >50K  
408            1887              40  United-States   >50K  
580               0              30  United-States   >50K  
...             ...             ...            ...    ...  
31997             0              12  United-States   >50K  
31998             0              40  United-States   >50K  
32016             0              40  United-States   >50K  
32530             0              55  United-States   >50K  
32539             0              10  United-States   >50K  

[191 rows x 15 columns]

### To observe fnlwgt vs income 

In [76]:
df[df['income'] == '>50K'][['fnlwgt', 'income']]


fnlwgt income
7      209642   >50K
8       45781   >50K
9      159449   >50K
10     280464   >50K
11     141297   >50K
...       ...    ...
48815  149347   >50K
48816   23157   >50K
48822  202168   >50K
48826  139347   >50K
48841  182148   >50K

[11687 rows x 2 columns]

In [80]:
df.groupby('income')['fnlwgt'].sum()


income
<=50K    7060920057
>50K     2202655605
Name: fnlwgt, dtype: int64

<h3>Notes: fnlwgt & income understanding </h3>
<span style =" color: #200FBD; font-size: 15px; font-family: Baskerville Old Face; font-weight: bold;">
<ol>
  <li> fnlwgt : How many people in given row represents in population </strong>.</li>
  <li> use of fnlwgt : Only for population - level weighing, not individual analysis. </strong></li>
  </ul>
  </li>
</ol>

### Data cleaning - Converting '?' with empty ''

In [111]:
df = df.replace('','Unknown')

df.loc[27]  # testing 27th index

age                               54
workclass                    Unknown
fnlwgt                        180211
education               Some-college
education-num                     10
marital-status    Married-civ-spouse
occupation                   Unknown
relationship                 Husband
race              Asian-Pac-Islander
sex                             Male
capital-gain                       0
capital-loss                       0
hours-per-week                    60
native-country                 South
income                          >50K
Name: 27, dtype: object

In [ ]:
### Unknown columns to track

In [112]:
unknown_cols = df.columns[(df== 'Unknown').any()].tolist()

for col in unknown_cols:
    print(f"Value counts for column '{col}':")
    print(df[col].value_counts())
    print("-" * 40)

Value counts for column 'workclass':
workclass
Private             33906
Self-emp-not-inc     3862
Local-gov            3136
State-gov            1981
Unknown              1836
Self-emp-inc         1695
Federal-gov          1432
Without-pay            21
Never-worked           10
Name: count, dtype: int64
----------------------------------------
Value counts for column 'occupation':
occupation
Prof-specialty       6172
Craft-repair         6112
Exec-managerial      6086
Adm-clerical         5611
Sales                5504
Other-service        4923
Machine-op-inspct    3022
Transport-moving     2355
Handlers-cleaners    2072
Unknown              1843
Farming-fishing      1490
Tech-support         1446
Protective-serv       983
Priv-house-serv       242
Armed-Forces           15
Name: count, dtype: int64
----------------------------------------
Value counts for column 'native-country':
native-country
United-States                 43832
Mexico                          951
Unknown          

In [102]:
df.occupation.value_counts()

occupation
Prof-specialty       6172
Craft-repair         6112
Exec-managerial      6086
Adm-clerical         5611
Sales                5504
Other-service        4923
Machine-op-inspct    3022
Transport-moving     2355
Handlers-cleaners    2072
Unknown              1843
Farming-fishing      1490
Tech-support         1446
Protective-serv       983
Priv-house-serv       242
Armed-Forces           15
Name: count, dtype: int64

<h3>Notes: One Hot Encoding </h3>
<span style =" color: #200FBD; font-size: 15px; font-family: Baskerville Old Face; font-weight: bold;">
<ol>
  <li> Converting to Binary forms -- categorical fields </strong>.</li>

  </ul>
  </li>
</ol>

#### to make one hot encoding with binary forms from occupation field

In [103]:
pd.get_dummies(df.occupation)

Adm-clerical  Armed-Forces  Craft-repair  Exec-managerial  \
0              True         False         False            False   
1             False         False         False             True   
2             False         False         False            False   
3             False         False         False            False   
4             False         False         False            False   
...             ...           ...           ...              ...   
48837         False         False         False            False   
48838         False         False         False            False   
48839         False         False         False            False   
48840          True         False         False            False   
48841         False         False         False             True   

       Farming-fishing  Handlers-cleaners  Machine-op-inspct  Other-service  \
0                False              False              False          False   
1                False              False              False          False   
2                False               True              False          False   
3                False               True              False          False   
4                False              False              False          False   
...                ...                ...                ...            ...   
48837            False              False              False          False   
48838            False              False              False          False   
48839            False              False              False          False   
48840            False              False              False          False   
48841            False              False              False          False   

       Priv-house-serv  Prof-specialty  Protective-serv  Sales  Tech-support  \
0                False           False            False  False         False   
1                False           False            False  False         False   
2                False           False            False  False         False   
3                False           False            False  False         False   
4                False            True            False  False         False   
...                ...             ...              ...    ...           ...   
48837            False            True            False  False         False   
48838            False           False            False  False         False   
48839            False            True            False  False         False   
48840            False           False            False  False         False   
48841            False           False            False  False         False   

       Transport-moving  Unknown  
0                 False    False  
1                 False    False  
2                 False    False  
3                 False    False  
4                 False    False  
...                 ...      ...  
48837             False    False  
48838             False    False  
48839             False    False  
48840             False    False  
48841             False    False  

[48842 rows x 15 columns]

In [104]:
occupation_dummies = pd.get_dummies(df['occupation']).astype(int)
occupation_dummies

Adm-clerical  Armed-Forces  Craft-repair  Exec-managerial  \
0                 1             0             0                0   
1                 0             0             0                1   
2                 0             0             0                0   
3                 0             0             0                0   
4                 0             0             0                0   
...             ...           ...           ...              ...   
48837             0             0             0                0   
48838             0             0             0                0   
48839             0             0             0                0   
48840             1             0             0                0   
48841             0             0             0                1   

       Farming-fishing  Handlers-cleaners  Machine-op-inspct  Other-service  \
0                    0                  0                  0              0   
1                    0                  0                  0              0   
2                    0                  1                  0              0   
3                    0                  1                  0              0   
4                    0                  0                  0              0   
...                ...                ...                ...            ...   
48837                0                  0                  0              0   
48838                0                  0                  0              0   
48839                0                  0                  0              0   
48840                0                  0                  0              0   
48841                0                  0                  0              0   

       Priv-house-serv  Prof-specialty  Protective-serv  Sales  Tech-support  \
0                    0               0                0      0             0   
1                    0               0                0      0             0   
2                    0               0                0      0             0   
3                    0               0                0      0             0   
4                    0               1                0      0             0   
...                ...             ...              ...    ...           ...   
48837                0               1                0      0             0   
48838                0               0                0      0             0   
48839                0               1                0      0             0   
48840                0               0                0      0             0   
48841                0               0                0      0             0   

       Transport-moving  Unknown  
0                     0        0  
1                     0        0  
2                     0        0  
3                     0        0  
4                     0        0  
...                 ...      ...  
48837                 0        0  
48838                 0        0  
48839                 0        0  
48840                 0        0  
48841                 0        0  

[48842 rows x 15 columns]